### ISSUE TITLE
error handling regression from 2.0.0-rc.0 to 2.0.0-rc.1

### ISSUE THREAD

**Issue by ctm** [diesel-rs/diesel#3282]
### Versions

- **Rust:** 1.63.0
- **Diesel:** 2.0.0-rc.1
- **Database:** Postgresql 12.8
- **Operating System** Linux de33d8ba93ea 5.4.0-1043-aws # 45~18.04.1-Ubuntu SMP Fri Apr 9 23:32:25 UTC 2021 x86_64 GNU/Linux

### Feature Flags

- **diesel:** ["postgres", "r2d2", "serde_json", "chrono"]

### Steps to reproduce

I apologize for not providing code, but I ran into this when running a closed source application that was connected to a Postgres 12.8 RDS instance after upgrading from 2.0.0-rc.0 to 2.0.0-rc.1. It does not happen using Postgres 13.8. Given time, I can probably write up a test case that demonstrates the issue, but a textual description of what happens is something I can write now.

Due to a bug, our app was trying to insert a value into a table in a way that violated the foreign key constraint on a column. Our code would detect the failure and use info! to log the problem. After switching from rc0 to rc1, we also started getting the above `r2d2` error after the foreign key constraint error. 

I did not notice the `r2d2` complaint until a few days ago and only today did I manage to use git bisect to see that it only started happening after our switch from rc0 to rc1 (which we did shortly after rc1 came out).

Based on the fact that it happens in Postgres V12 but not V13 and looking at the actual change to the r2d2 code, my guess is that in V12, errors had to be explicitly cleared by the client but now they don't have to be.

I'm going to upgrade our RDS instance to whatever the most recent Postgres version is available and I'm also fixing the bug that causes the foreign key violation, but I figure it makes sense to report this bug even though it won't ever bite us again.

## Checklist

- [x] I have already looked over the [issue tracker](https://github.com/diesel-rs/diesel/issues)  and the [disussion forum](https://github.com/diesel-rs/diesel/discussions)  for similar possible closed issues.
 
- [x] This issue can be reproduced on Rust's stable channel. (Your issue will be
  closed if this is not the case)
- [ ] This issue can be reproduced without requiring a third party crate (I have not yet verified this, but I strongly suspect it)

**Comment by weiznich** [comment]
This is likely causes by https://github.com/diesel-rs/diesel/pull/3213. We would need an reproducible example for finding the exact cause for this issue, as the relevant PR is quite large.

**Comment by ctm** [comment]
Thank you for the quick response. I apologize for not providing example code yet.

My take is completely naive and so you're probably right, but my guess was that the problem was due to:
```
 #[cfg(feature = "r2d2")]
 impl crate::r2d2::R2D2Connection for PgConnection {
     fn ping(&mut self) -> QueryResult<()> {
@@ -177,11 +280,20 @@ impl crate::r2d2::R2D2Connection for PgConnection {
     }
 
     fn is_broken(&mut self) -> bool {
-        self.transaction_state
+        match self
+            .connection_and_transaction_manager
+            .transaction_state
             .status
             .transaction_depth()
-            .map(|d| d.is_none())
-            .unwrap_or(true)
+        {
+            // all transactions are closed
+            // so we don't consider this connection broken
+            Ok(None) => false,
+            // The transaction manager is in an error state
+            // or contains an open transaction
+            // Therefore we consider this connection broken
+            Err(_) | Ok(Some(_)) => true,
+        }
     }
 }
```
and that the issue was simply that under Postgres V12, the connection was left in an error state and that Postgres changed that surprising behavior. This guess of mine is based on the fact that 2.0.0-rc.1 works fine with a more recent version of Postgres.

I do plan to get you a small example, but I'm trying to meet a (admittedly self-imposed) deadline to get some of our stuff working (including an in-progress migration from Postgres V12 to Postgres V14) in the next few hours, simply because that's when the testers will appear for the day.

**Comment by weiznich** [comment]
Fixed by #3344
